src/_pytest/recwarn.py

```diff
@@ -303,6 +323,8 @@ def found_str():
                     for r in self:
                         if issubclass(r.category, self.expected_warning):
                             if re.compile(self.match_expr).search(str(r.message)):
+                                if self.re_emit:
```

Comment:
the logic here is slightly terrifying at the nesting level, at first glance i had the impression this must be wrong

however was we break right after it its indeed correct